### PR TITLE
[CPDLP-2834] initial models for Statement and StatementItem

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net/browse/CPDLP-2834

### State machine library options

| Library  | Latest Release | Ruby Toolbox Score |
| ------------- | ------------- | ------------- |
| [State Machines](https://github.com/state-machines/state_machines)  | 2023-06-30 | 4.04 |
| [Acts as state machine (AASM)](https://github.com/aasm/aasm) | 2023-02-05 | 3.77 |
| [Statesman](https://github.com/gocardless/statesman) | 2024-01-09 | 0.80 |

### Changes proposed in this pull request

* `Statement` model
* `StatementItem` model
* Proposed state implementation with `state_machines-activerecord` gem
* Added `docs/state_machines.md`, a place for state diagrams

### CODE REVIEW

File: app/models/statement.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  validates :output_fee, presence: true
+  validates :month, numericality: { in: 1..12, only_integer: true }
+  validates :year, numericality: { in: 2020..2050, only_integer: true }
+
+  validate :validate_max_statement_items_count
+
+  state_machine :state, initial: :open do
+    state :open
+    state :payable
+    state :paid
```
Comment (peteryates) [inline app/models/statement.rb:15]: So is it definitely correct to say that every statement ever takes this route? None are cancelled or closed any other way?

File: db/migrate/20240221113542_create_statements.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      t.integer :month, null: false
+      t.integer :year, null: false
+      t.date :deadline_date
+      t.date :payment_date
+      t.boolean :output_fee, default: true, null: false
+      t.references :cohort, null: false, foreign_key: true
+      t.references :lead_provider, null: false, foreign_key: true
+      t.datetime :marked_as_paid_at
+      t.decimal :reconcile_amount, precision: 8, scale: 2
+      t.datetime :frozen_at
```
Comment (peteryates) [inline db/migrate/20240221113542_create_statements.rb:15]: What does frozen mean here? Can `open`, `payable` and `paid` statements all be frozen?

File: docs/state_machines.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+## Statement
+
+```mermaid
+---
+title: State
+---
+stateDiagram-v2
+    [*] --> open
+    open --> payable: mark_payable
+    payable --> paid: mark_paid
```
Comment (peteryates) [inline docs/state_machines.md:12]: I like this.

I think we can automate it. Looking at the gem we can get a list of all possible states like this:

```ruby
Statement.state_machines[:state].states.map(&:name)
=> [:open, :payable, :paid]
```

Then we can loop through and do something [similar to the official gem](https://github.com/state-machines/state_machines-graphviz/blob/master/lib/state_machines/graphviz/monkeypatch.rb) but build up Mermaid transitions.

Could even be a rake task where we find and replace in this file.

**Comment by peteryates** [comment]
Looks good to me overall. Just left a couple of small questions on the `Statement` states. To the uninitiated, it looks like `frozen` is a state.

File: db/migrate/20240221113542_create_statements.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    create_table :statements do |t|
+      t.integer :month, null: false
+      t.integer :year, null: false
+      t.date :deadline_date
+      t.date :payment_date
+      t.boolean :output_fee, default: true, null: false
+      t.references :cohort, null: false, foreign_key: true
+      t.references :lead_provider, null: false, foreign_key: true
+      t.datetime :marked_as_paid_at
```
Comment (peteryates) [inline db/migrate/20240221113542_create_statements.rb:13]: Another question to the uninitiated, what's the difference between `payment_date` and `marked_as_paid_at` (other than it being a `datetime`)

File: db/migrate/20240221113542_create_statements.rb (hunk 2 above)
Comment (mooktakim) [reply db/migrate/20240221113542_create_statements.rb:15]: @peteryates 
`frozen_at` is a new field added to the separation branch: https://github.com/DFE-Digital/npq-registration/commit/4380f4ed1973090f03597aa56d2a4e3c8a4eff18

I'm not sure if we need it, the statement should be "frozen" once its paid.

File: app/models/statement.rb (hunk 1 above)
Comment (mooktakim) [reply app/models/statement.rb:15]: I don't think there's ever been a cancelled statement. I believe there is always a statement for each month.
Once the statement is `paid` its no longer added to, and shouldn't change.

File: db/migrate/20240221113542_create_statements.rb (hunk 4 above)
Comment (mooktakim) [reply db/migrate/20240221113542_create_statements.rb:13]: `marked_as_paid_at` when we action statement to be paid, it is used to display some text and for background job:
https://dfedigital.atlassian.net/browse/CPDLP-2295

File: db/migrate/20240221113542_create_statements.rb (hunk 2 above)
Comment (peteryates) [reply db/migrate/20240221113542_create_statements.rb:15]: Ah, perhaps @pmanrubia or @cwrw can shed some light.

File: db/migrate/20240221113542_create_statements.rb (hunk 4 above)
Comment (peteryates) [reply db/migrate/20240221113542_create_statements.rb:13]: Ah ok, so:

* `payment_date` is the date on which the payment _was_ made
* `marked_as_paid_at` is the date and time *we recorded it* as being paid

File: app/models/statement.rb (hunk 1 above)
Comment (peteryates) [reply app/models/statement.rb:15]: Ok, makes sense - thanks!

File: db/migrate/20240221113542_create_statements.rb (hunk 2 above)
Comment (cwrw) [reply db/migrate/20240221113542_create_statements.rb:15]: Having a look at the comment here https://github.com/DFE-Digital/npq-registration/pull/1036 basically in ECF currently we use the type "Payable" to allow freezing a statement in our logic. I think we can probably rely on deadline date or the enum state, so probably we can drop this

File: app/models/statement_item.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+class StatementItem < ApplicationRecord
```
Comment (cwrw) [inline app/models/statement_item.rb:1]: curious around the naming decision (I guess that's what we landed on with the analysis) vs line item?

**Review by cwrw — COMMENTED** [review]
very nicely done 🙌  just the comment around frozen at otherwise 👍

File: app/models/statement_item.rb (hunk 5 above)
Comment (peteryates) [reply app/models/statement_item.rb:1]: I'm not massively invested in either line or item but they both mean the same thing here so using both doesn't feel like it makes sense.

Comment (mooktakim) [reply app/models/statement_item.rb:1]: In the analysis, we talked about this just being a shorter maybe name.

But I wonder, maybe we should instead call it `StatementDeclaration`, as it closely describes the many-to-many relationship.

File: db/migrate/20240221113542_create_statements.rb (hunk 2 above)
Comment (mooktakim) [reply db/migrate/20240221113542_create_statements.rb:15]: I've removed `frozen_at` for now, if we need it, we can add.

**Review by cwrw — APPROVED** [review]

File: app/models/statement_item.rb (hunk 5 above)
Comment (cwrw) [reply app/models/statement_item.rb:1]: I think that sounds reasonable, the only thing that makes me hesitant is that there is a `state` on the record, so happy to stick to items for now.

**Pull request merged**